### PR TITLE
v2.2.1

### PR DESCRIPTION
**Pull request by debsahu**
* checkForRequests() is vital for e131 mode, remove from #ifdef ENABLE_LEGACY_ANIMATIONS
 * Minor fixes related to NeoPixelBus UART methods
 * Modify platformio.ini for future bump to esp8266-arduino v2.5.0 (shamelessly stolen settings from espurna project)
 * Gzipped index2.htm & edit.htm.gz(untouched), converted to hex format using xxd -i abcd.gz > html_gz.h
 * Pointers added for WS2812FX & NeoPixelBus
 * new "REST API": /pixels?ct=xxx to change length of LED strip
 * new "REST API": /pixels?rgbo=xxx to change RGB order
 * new "REST API": /pixels?pin=GPIO_NO to change PIN# (Allowed GPIO values: 16/5/4/0/2/14/12/13/15/3/1)
 * added HA 0.87 version support https://github.com/toblum/McLighting/issues/327
 * Added alternative way to send large messages using PubSubClient
 * Bump PIO core to 2.0.1
 * Send HA state on MQTT connect, address https://github.com/toblum/McLighting/issues/349
 * Add LWT for MQTT and AMQTT, address https://github.com/toblum/McLighting/issues/340
 * Added file for custom WS2812FX animations in custom slots
 * Rename variables to be char instead of String
 * Added LED pixel count and PIN settings to WiFiManager
 * autoParams now uses milliseconds instead of seconds

### CODE REVIEW

**Comment by debsahu** [comment]
I tested all of the new features, everything works as of now.

**Comment by M4lik** [comment]
As far as I am concerned with the general API structure this seems fine to me.
Usually I would make these API request via a POST and not a GET request, but this should be a task for another day. Tomorrow I will start to refactor a little bit of the API code and we will see how well this goes 😅

File: Arduino/McLighting/McLighting.ino [hunk 1]
```diff
@@ -692,6 +771,93 @@ void setup() {
     getStatusJSON();
   });
 
+  server.on("/pixels", []() {
```
Comment (M4lik) [inline Arduino/McLighting/McLighting.ino]: As said, this looks fine to me.

File: Arduino/McLighting/McLighting.ino [hunk 2]
```diff
@@ -692,6 +771,93 @@ void setup() {
...
+      if (pixelCt > 0) {
+        WS2812FXStripSettings.stripSize = pixelCt;
+        updateStrip = true;
+        DBG_OUTPUT_PORT.printf("/pixels: Count# %d\n", pixelCt);
+      }
+    }
+    if(server.hasArg("rgbo")){
+      String RGBOrder = server.arg("rgbo");
+      DBG_OUTPUT_PORT.print("/pixels: RGB Order# ");
+      if (RGBOrder == "grb") {
```
Comment (M4lik) [inline Arduino/McLighting/McLighting.ino]: I'm not an programming expert in Arduino/c/c++, but couldn't this be better solved via an extra `if` on the very bottom of this function, where we set
```
updateStrip = true;
DBG_OUTPUT_PORT.println(RGBOrder);
```
if  this if-else-ladder did not fail?
Would make this a little bit more cleaner.
But as I said: Not an expert.

File: Arduino/McLighting/McLighting.ino (hunk 1 above)
Comment (M4lik) [reply Arduino/McLighting/McLighting.ino]: One more addition: I like it, that you put all three operations in one endpoint, but I think `pixels` is a bit misleading. Maybe change it to `pixelconf`?

File: Arduino/McLighting/McLighting.ino [hunk 3]
```diff
@@ -692,6 +771,93 @@ void setup() {
...
+        WS2812FXStripSettings.pin = pin;
+        updateStrip = true;
+        DBG_OUTPUT_PORT.println(pin);
+      } else {
+        DBG_OUTPUT_PORT.println("invalid input!");
+      }
+      #endif
+    }
+
+    if(updateStrip)
```
Comment (M4lik) [inline Arduino/McLighting/McLighting.ino]: Do we really need to check this? I mean: If nothing changed, we still could call these few lines. Nothing would happen, as nothing has changed...

File: Arduino/McLighting/McLighting.ino (hunk 1 above)
Comment (debsahu) [reply Arduino/McLighting/McLighting.ino]: We can use anything here, I just went with what came to my thoughts at that moment. I like the idea of using `pixelconf`. We can add the HTTP method later, restricting to only POST when the web interface to set these up is complete. That way users have a way to communicate and set pixel configuration.

File: Arduino/McLighting/McLighting.ino (hunk 2 above)
Comment (debsahu) [reply Arduino/McLighting/McLighting.ino]: I'm still learning, no one is an expert. What I found is new webserver code (core 2.5.2) sends each argument separately, that's why I used a Boolean as a quick check. We also do validation of each input received.

File: Arduino/McLighting/McLighting.ino (hunk 3 above)
Comment (debsahu) [reply Arduino/McLighting/McLighting.ino]: One never know what input one can receive. Always picture the user as a 4 yr old using a remote! (Ha ha) Let's say someone accidently uses GPIO10, one can fry the Flash. I did it in espeasy setup, that why I know! The web interface allowed any pin to be configured without these checks and magic smoke!

File: Arduino/McLighting/McLighting.ino (hunk 1 above)
Comment (debsahu) [reply Arduino/McLighting/McLighting.ino]: Changed to `pixelconf` in https://github.com/toblum/McLighting/pull/350/commits/1f0087fec856708075e9fe087d92b4d0dc8001d6

File: Arduino/McLighting/McLighting.ino (hunk 2 above)
Comment (debsahu) [reply Arduino/McLighting/McLighting.ino]: Having one bool do the job is better than multiple if statement checks. **updateStrip** does valid input check for multiple inputs, so acts as an "OR" for now.

**Comment by debsahu** [comment]
> As far as I am concerned with the general API structure this seems fine to me.
> Usually I would make these API request via a POST and not a GET request, but this should be a task for another day. Tomorrow I will start to refactor a little bit of the API code and we will see how well this goes 😅

https://github.com/toblum/McLighting/pull/350/commits/3af70b4953a6265c79a2dd99cb6b0a607bc0eb85 has the HTTP method check commented, but can be used in the future when web-interface is ready.

**Comment by debsahu** [comment]
@toblum Do you want `#define USE_HTML_MIN_GZ` commented or un-commented? This is gzipped version of the web-interface so the user doesn't have to upload web-page on SPIFFs. We can also release binaries with this release, because user can change LED settings.

File: Arduino/McLighting/McLighting.ino [hunk 4]
```diff
@@ -692,6 +771,105 @@ void setup() {
...
+        updateStrip = true;
+        DBG_OUTPUT_PORT.println(RGBOrder);
+      } else {
+        DBG_OUTPUT_PORT.println("invalid input!");
+      }
+    }
+    if(server.hasArg("pin")){
+      uint8_t pin = server.arg("pin").toInt();
+      DBG_OUTPUT_PORT.print("/pixels: GPIO used# ");
+      #if defined(USE_WS2812FX_DMA) or defined(USE_WS2812FX_UART1) or defined(USE_WS2812FX_UART2)
```
Comment (M4lik) [inline Arduino/McLighting/McLighting.ino]: One more thing... Why the outer `if` check? You check `USE_WS2812FX_DMA`, `USE_WS2812FX_UART1` and `USE_WS2812FX_UART2` twice....

File: Arduino/McLighting/definitions.h [hunk 5]
```diff
@@ -52,35 +55,36 @@ const char HOSTNAME[] = "McLighting01";   // Friedly hostname
 #endif
 
 // parameters for automatically cycling favorite patterns
-uint32_t autoParams[][4] = { // color, speed, mode, duration (seconds)
-  {0xff0000, 200,  1,  5.0}, // blink red for 5 seconds
-  {0x00ff00, 200,  3, 10.0}, // wipe green for 10 seconds
-  {0x0000ff, 200, 11,  5.0}, // dual scan blue for 5 seconds
-  {0x0000ff, 200, 42, 15.0}  // fireworks for 15 seconds
+uint32_t autoParams[][4] = {  // color, speed, mode, duration (milliseconds)
```
Comment (M4lik) [inline Arduino/McLighting/definitions.h]: Verry much approve this. :+1:

Comment (debsahu) [reply Arduino/McLighting/definitions.h]: Contribution by @dale3h https://github.com/toblum/McLighting/commit/67a98c28b899dbb608bd32da291e3e250e1f41ff

File: Arduino/McLighting/McLighting.ino (hunk 4 above)
Comment (debsahu) [reply Arduino/McLighting/McLighting.ino]: They are just pre-compiler flags. If one comes one of those methods during compilation, user cannot change the GPIO pin as those methods are hardware limited to those pins on ESP8266. In the future we should be able to switch between these methods, but for now these portions of the code are under pre-processor flags.

Comment (debsahu) [reply Arduino/McLighting/McLighting.ino]: TODO: Make switching between methods: Adafruit NeoPixel, NeoPixelBus's DMA, NeoPixelBus's UART1 and NeoPixelBus's UART2 as REST API.

Comment (debsahu) [reply Arduino/McLighting/McLighting.ino]: Added to todo.

**Comment by toblum** [comment]
Hi @debsahu,

this new feature packed version lloks great. Unfortunately I didn't have the possibility to check this out yet with the new bundeled interface. I just did some short tests last weekend with my usual setup ist the dev version worked great.
The only concern I have with the option enabled by default is that people might get confused if theyy follow the "old" instructions from the videos. But we'll see. Let's give it a try.
Having the interface packed makes setup much faster, that outweighs the disadvantages by far.
I hope that I'm back at the bench at the weekend and will try that version out.
I was thinking about creating a completely "could-based" UI that works as a progressive web app that you can "install" on your phone and that shoulld work also offline. I have to see if browser support is as far as I expect it to be. We'll see. That new config options via API come in handy for that.

Greetings Tobias

**Comment by debsahu** [comment]
@toblum It will not matter even if they upload the file on SPIFFs, it will still server the PROGMEM version only for index.htm. Other files are served from SPIFFs.

The web-interface is where my skills lack, but I'm learning. You expertise in that front would be great, but there is no hurry.

**Comment by toblum** [comment]
Oh, I see. I didn't recognize this from the source. I'll have to try that out by myself. That makes it even more flexible. Perfect!

**Pull request merged**